add a mock CLI command for calling Apalache (#787)
informalsystems/quint: doc/quint.md

   The errors and warnings are written in the format of [ADR002][].

## Command verify

```sh
$ quint verify <input>

Verify a Quint specification via Apalache

Options:
  --help             Show help                                     [boolean]
  --version          Show version number                           [boolean]

>   --how              how to do verification: bounded model checking (bmc) or

Reviewer: In our meeting today we discussed that maybe calling this `strategy` would be clearer? 

But for the sake of giving the simplest API for prototyping here, as we also discussed, I propose simply removing this flag. In apalache, "simulate" is must just the check command with two tuning parameter set:

https://github.com/informalsystems/apalache/blob/2cec6c7be5d66e91ee7653928dfa2b61269aa0b9/mod-tool/src/main/scala/at/forsyte/apalache/tla/tooling/opt/SimulateCmd.scala#L5-L16

We can set these via the config inputs currently, which can give us time to iterate on a more thought out interface after using it. WDYT?

>   --how              how to do verification: bounded model checking (bmc) or

Author: If it makes it easier, we can keep `check` only. Though as far as I understand, it is just mapping one set of flags to another.

                     symbolic execution (symexec)

>                    [string] [choices: "symexec", "bmc"] [default: "symexec"]

Author: Why do you think that `apalache simulate` is more suitable under `run`? Both `check` and `simulate` are symbolic techniques. Indeed, they have different kinds of guarantees, but they are both incomplete from the point of view of complete verification.

>                    [string] [choices: "symexec", "bmc"] [default: "symexec"]

Reviewer: The fact that a subset of the flags work with one value for this flag but not another is a good indication that they should be separate commands IMO.

But, also, form a UX perspective, whether or not they are both symbolic techniques doesn't seem like a good grounds for grouping them together? Shouldn't the question be: are the used for the same thing?

>                    [string] [choices: "symexec", "bmc"] [default: "symexec"]

Author: I think `check` and `simulate` are used interchangeably in Apalache. In this sense, they are mostly compatible with one another. It's true that `simulate` has a bit more flags.

  --main             name of the main module (by default, computed from
                     filename)                                      [string]
  --out              output file (suppresses all console output)    [string]
  --out-itf          output the trace in the Informal Trace Format to file
                     (supresses all console output)                 [string]
  --max-samples      the number of symbolic paths to try (when using
                     --how=symexec)                [number] [default: 10000]
  --max-steps        the maximum on the number of steps in every trace
                                                      [number] [default: 10]
  --init             name of the initializer action
                                                  [string] [default: "init"]
  --step             name of the step action      [string] [default: "step"]
  --invariant        invariant to check: a definition name or an expression
                                                [string] [default: ["true"]]
  --seed             random seed to use for symbolic execution (when using
                     --how=symexec)                                 [string]

>   --apalache-config  Filename of the additional Apalache flags in the JSON
>                      format                                         [string]

Reviewer: ```suggestion
  --apalache-config  Filename of the additional Apalache configuration in the HOCON
                     format (a superset of JSON)                           [string]
```

Author: Fix in cli.ts, since this help is autogenerated

```

- If there are no critical errors (e.g., in parsing, typechecking, etc.), this
command sends the Quint specification to the [Apalache][] model checker, which
will try to find an invariant violation. If it finds one, it prints the trace
on the standard output. If it does not find a violating trace, it prints the

> longest sample trace that the simulator has found during the execution. When the

Reviewer: This sentence only applies to one mode. To make it accurate we need a conditional here, which I guess is another indication that theses should be factored into separate commands? :)

parameter `--out` is supplied, the trace is written as a JSON representation of

> Quint IR in the output file. When the parameter `--out-itf` is supplied, the

Author: This is an interesting idea. Would be easy to discover the ITF feature though?

> Quint IR in the output file. When the parameter `--out-itf` is supplied, the

Reviewer: I think it would be easy to discover if documented? :)

> Quint IR in the output file. When the parameter `--out-itf` is supplied, the

Reviewer: actually, a more flexible approach might be to have a separate flag that specifies the output format. This would let us also output to stdout via something like `quint foo --out - --fmt (itf|json)`?

> Quint IR in the output file. When the parameter `--out-itf` is supplied, the

Reviewer: thanks to some refactoring work, i'm reminded that `out` is a very specific kind of data really. Maybe we should consider giving a more particular name to reflect that? In any case, overloading it with trace output when it's used currently for compilation data doesn't seem like a very good idea. This answers my initial quetsion :D

trace is written in the [Informal Trace Format][]. This output can be conviently
displayed with the [ITF Trace Viewer][], or just with [jq][].

  Apalache implements two slightly different techniques for invariant verification:
    
    - Symbolic execution (`symexec`). This technique selects `--max-samples`
    runs in the specification and executes them symbolically via the [z3][]

>     prover. The length of every run is limited with `--max-steps` steps. This
>     techique is usually finding bugs faster than `bmc`, but it does check only
>     *some runs* up to `--max-steps` steps.

Reviewer: ```suggestion
    prover. The length of every run is limited with `--max-steps` steps. This
    techique is can find bugs faster than `bmc`, but it only checks *some 
    runs* up to `--max-steps` steps.
```

    
    - Bounded model checking (`bmc`). This technique checks *all runs* up to
    `--max-steps` steps via [z3][]. It is usually slower than `symexec` at bug
    finding. However, when it terminates and does not find an invariant

>     violation, it comes with a guarantee: The invariant cannot be violated by

Reviewer: The guarantee vs. lack thereof is, imo, the obvious answer to your question :)

> Why do you think that apalache simulate is more suitable under run

A check that might or might not find some bugs (within the given constraints) seems less like a "verification" and much more like a test?
